docs/proposals/propagation.md

```diff
@@ -0,0 +1,142 @@
+sharing the mount propagation.
+
+1. (From @YorikSar) For OpenStack project, Neutron, we need network namespaces
+created by it to persist across reboot of pods with Neutron agents. Without it
+we have unnecessary data plane downtime during rolling update of these agents.
+Neutron L3 agent creates interfaces and iptables rules for each virtual router
+in a separate network namespace. For managing them it uses ip netns command that
+creates persistent network namespaces by calling unshare(CLONE_NEWNET) and then
+bind-mounting new network namespace's inode from /proc/self/ns/net to file with
+specified name in /run/netns dir. These bind mounts are the only references to
+these namespaces that remain.
+When we restart the pod, its mount namespace is destroyed with all these bind
+mounts, so all network namespaces created by the agent are gone. For them to
+survive we need to bind mount a dir from host mount namespace to container one
+with shared flag, so that all bind mounts are propagated across mount namespaces
+and references to network namespaces persist.
+
+
+## Implementation Alternatives
+
+### Add an option in VolumeMount API
+
+The new `VolumeMount` will look like:
+```go
+type VolumeMount struct {
+	// Required: This must match the Name of a Volume [above].
+	Name string `json:"name"`
+	// Optional: Defaults to false (read-write).
+	ReadOnly bool `json:"readOnly,omitempty"`
+	// Required.
+	MountPath string `json:"mountPath"`
+	// Optional.
+	Propagation string `json:"propagation"`
+}
+```
+
+Opinion against this:
+
+1. This will affect all volumes, while only HostPath need this.
+
+1. This need API change, which is discouraged.
+
+### Add an option in HostPathVolumeSource
+
+The new `HostPathVolumeSource` will look like:
+```go
+const (
+	PropagationShared  PropagationMode = "Shared"
+	PropagationSlave   PropagationMode = "Slave"
+	PropagationPrivate PropagationMode = "Private"
+)
+
+type HostPathVolumeSource struct {
+	Path string `json:"path"`
+	// Mount the host path with propagation mode specified. Docker only.
+	Propagation PropagationMode `json:"propagation,omitempty"`
+}
+```
+
+Opinion against this:
+
+1. This need API change, which is discouraged.
+
+1. All containers use this volume will share the same propagation mode.
+
+### Make shared default for HostPath
+
+Given only HostPath needs this feature, and privileged access is needed when
+making mounts inside container, we can bind propagation mode with existing option
+privileged.
+
+The propagation mode could be determined by the following logic:
+```go
+// Environment check to ensure "rshared" is supported.
+if !dockerNewerThanV110 || !mountPathIsShared {
+	return ""
+}
+if container.SecurityContext.Privileged {
+	return "rshared"
+} else {
+	return "rslave"
+}
+```
+
+Opinion against this:
+
+1. This changes the behavior of existing config.
+
+1. "shared" is not supported by some kernels, we need runtime support matrix
```

Comment:
cc @pmorie, this is from your comment on the previous issue, if you have something to reference that would be great!

@lvlv if you're up for doing the legwork to figure out how bad this is, it would be good to know. At the very least, we'll need to document the requirements to use this feature.